README

Picsnvideos is a plugin for JPilot. It fetches pictures and
videos from late model Palm smartphones.

Picsnvideos supports the Treo 650, Treo 680, Treo 700p, Treo 755p, and
the Centro.  Only the Treo 650 and Centro have been known to be tested.
Please post your experiences with untested devices at 
http://sourceforge.net/projects/picsnvideos .

This plugin fetches, rather than syncs.  The plugin reads pictures and
videos taken with the camera on the Palm, and stores them on the computer.  
It will also fetch audio captions added with the 'Pics&Videos'
application (known as 'Media' on older Palms). Files are never deleted
from the Palm or from the computer.

Files from the internal memory of the Palm are stored in 
'$HOME/PalmPictures/Device'.  Files from the Palm's SD Card are stored in 
'$HOME/PalmPictures/SDCard'.

Each album on the Palm is a directory under 'Device/' or 'SDCard/'.
Some older Palms will create an 'Unfiled/' directory for pictures and
videos that have not yet been moved to an album.

The file extensions of the files that picsnvideos fetches include:
    .jpg - JPEG pictures
    .amr - audio photo captions from GSM phones
    .qcp - audio photo captions from CDMA phones
    .3gp - video clips from GSM phones
    .3g2 - video clips from CDMA phones

The audio photo captions are named <photo_name.jpg>.amr or 
<photo_name.jpg>.qcp.  For example, the picture 
photo_051608_001.jpg  will have an audio caption named 
photo_051608_001.jpg.amr (or .qcp).

Once a picture or video has been fetched, it will never be fetched
again even if it moved to a different album.  Re-recorded audio captions
will be refetched.  To force all pictures to be re-fetched, delete
the file $HOME/.jpilot/picsnvideos-fetched.gdbm.

Problems or suggestions can be reported in the forums or tracker at
http://sourceforge.net/projects/picsnvideos.  it is helpful to include 
the output that 'jpilot -d' creates whey you sync.
This output goes both to standard output and to 'jpilot.log'.


NEWS
(this file intentionally left blank)


libplugin.h
/* $Id: libplugin.h,v 1.1.1.1 2008/05/15 00:16:27 danbodoh Exp $ */

/*******************************************************************************
 * libplugin.h
 * A module of J-Pilot http://jpilot.org
 *
 * Copyright (C) 1999-2002 by Judd Montgomery
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 ******************************************************************************/

#ifndef __LIBPLUGIN_H__
#define __LIBPLUGIN_H__

/* #include "config.h" removed by Dan Bodoh, use when we set up configure */
/* #include <gtk/gtk.h> commented out by Dan Bodoh, gtk not used here */
#include <time.h>
#include <pi-appinfo.h>

/*
 * PLUGIN API for J-Pilot
 */

#ifdef ENABLE_PROMETHEON
#define PN "CoPilot"
#else
#define PN "J-Pilot"
#endif

#ifdef ENABLE_PROMETHEON
#define EPN "copilot"
#else
#define EPN "jpilot"
#endif

/*
 * For versioning of files
 */
#define FILE_VERSION     "version"
#define FILE_VERSION2    "version2"
#define FILE_VERSION2_CR "version2\n"


typedef struct {
  unsigned char Offset[4];  /*4 bytes offset from BOF to record */
  unsigned char attrib;
  unsigned char unique_ID[3];
} record_header;

typedef struct {
   unsigned long header_len;
   unsigned long header_version;
   unsigned long rec_len;
   unsigned long unique_id;
   unsigned long rt; /* Record Type */
   unsigned char attrib;
} PC3RecordHeader;

typedef struct mem_rec_header_s {
   unsigned int rec_num;
   unsigned int offset;
   unsigned int unique_id;
   unsigned char attrib;
   struct mem_rec_header_s *next;
} mem_rec_header;

#define LEN_RAW_DB_HEADER 78

typedef struct {
   char db_name[32];
   unsigned int flags;
   unsigned int version;
   time_t creation_time;
   time_t modification_time;
   time_t backup_time;
   unsigned int modification_number;
   unsigned int app_info_offset;
   unsigned int sort_info_offset;
   char type[5];/*Database ID */
   char creator_id[5];/*Application ID */
   char unique_id_seed[5];
   unsigned int next_record_list_id;
   unsigned int number_of_records;
} DBHeader;

int get_next_unique_pc_id(unsigned int *next_unique_id);

/* used for jp_delete_record */
#define CLEAR_FLAG    1
#define CANCEL_FLAG   2
#define DELETE_FLAG   3
#define MODIFY_FLAG   4
#define NEW_FLAG      5
#define COPY_FLAG     6
#define UNDELETE_FLAG 7

#define CLIST_DEL_RED 0xCCCC
#define CLIST_DEL_GREEN 0xCCCC
#define CLIST_DEL_BLUE 0xCCCC
#define CLIST_NEW_RED 55000
#define CLIST_NEW_GREEN 55000
#define CLIST_NEW_BLUE 65535
#define CLIST_MOD_RED 55000
#define CLIST_MOD_GREEN 65535
#define CLIST_MOD_BLUE 65535
#define CLIST_PRIVATE_RED 60000
#define CLIST_PRIVATE_GREEN 55000
#define CLIST_PRIVATE_BLUE 55000
#define CLIST_OVERDUE_RED 0xD900
#define CLIST_OVERDUE_GREEN 0x0000
#define CLIST_OVERDUE_BLUE 0x0000


#define DIALOG_SAID_1        454
#define DIALOG_SAID_PRINT    454
#define DIALOG_SAID_FOURTH   454
#define DIALOG_SAID_CURRENT  454
#define DIALOG_SAID_2        455
#define DIALOG_SAID_LAST     455
#define DIALOG_SAID_ALL      455
#define DIALOG_SAID_3        456
#define DIALOG_SAID_CANCEL   456

#define JP_LOG_DEBUG  1    /*debugging info for programmers, and bug reports */
#define JP_LOG_INFO   2    /*info, and misc messages */
#define JP_LOG_WARN   4    /*worse messages */
#define JP_LOG_FATAL  8    /*even worse messages */
#define JP_LOG_STDOUT 256  /*messages always go to stdout */
#define JP_LOG_FILE   512  /*messages always go to the log file */
#define JP_LOG_GUI    1024 /*messages always go to the gui window */

#define JPILOT_EOF -7

extern int jp_logf(int level, char *format, ...);

/* This bit means that this record is of no importance anymore */
#define SPENT_PC_RECORD_BIT 256

typedef enum {
   PALM_REC = 100L,
   MODIFIED_PALM_REC = 101L,
   DELETED_PALM_REC = 102L,
   NEW_PC_REC = 103L,
   DELETED_PC_REC =  SPENT_PC_RECORD_BIT | 104L,
   DELETED_DELETED_PALM_REC =  SPENT_PC_RECORD_BIT | 105L,
   REPLACEMENT_PALM_REC = 106L
} PCRecType;

typedef struct
{
   PCRecType rt;
   unsigned int unique_id;
   unsigned char attrib;
   void *buf;
   int size;
} buf_rec;

typedef struct
{
   char *base_dir;
   int *major_version;
   int *minor_version;
} jp_startup_info;

struct search_result
{
   char *line;
   unsigned int unique_id;
   struct search_result *next;
};

int plugin_get_name(char *name, int len);
int plugin_get_menu_name(char *name, int len);
int plugin_get_db_name(char *db_name, int len);
int plugin_startup(jp_startup_info *info);
/* int plugin_gui(GtkWidget *vbox, GtkWidget *hbox, unsigned int unique_id);
 *    commented out by Dan Bodoh; not using gtk
 */
int plugin_help(char **text, int *width, int *height);
int plugin_gui_cleanup(void);
int plugin_pre_sync_pre_connect(void);
int plugin_pre_sync(void);
int plugin_sync(int sd);
int plugin_search(const char *search_string, int case_sense, struct search_result **sr);
int plugin_post_sync(void);
int plugin_exit_cleanup(void);
int plugin_unpack_cai_from_ai(struct CategoryAppInfo *cai,
			      unsigned char *ai_raw, int len);
int plugin_pack_cai_into_ai(struct CategoryAppInfo *cai,
			    unsigned char *ai_raw, int len);
/* callbacks are needed for print */

void jp_init();
extern FILE *jp_open_home_file(char *filename, char *mode);

/* This takes the value of $JPILOT_HOME and appends /.jpilot/ and {file}
 * onto it and puts it into full_name.  max_size is the size if the
 * supplied buffer full_name
 */
int jp_get_home_file_name(char *file, char *full_name, int max_size);

/*
 * DB_name should be without filename ext, e.g. MemoDB
 * bufp is the packed app info block
 * size_in is the size of bufp
 */
int jp_pdb_file_write_app_block(char *DB_name, void *bufp, int size_in);

/*
 * widget is a widget inside the main window used to get main window handle
 * db_name should be without filename ext, e.g. MemoDB
 * cai is the category app info.  This should be unpacked by the user since
 * category unpack functions are database specific.
 */
/* commented out by Dan Bodoh, not using gtk
int jp_edit_cats(GtkWidget *widget, char *db_name, struct CategoryAppInfo *cai);
*/

/*************************************
 * convert char code
 *************************************/
extern void jp_charset_j2p(char *buf, int max_len);
extern void jp_charset_p2j(char *buf, int max_len);

/* file must not be open elsewhere when this is called, the first line is 0 */
int jp_install_remove_line(int deleted_line);

int jp_install_append_line(char *line);

/*
 * Get the application info block
 */
int jp_get_app_info(char *DB_name, unsigned char **buf, int *buf_size);
/*
 * Read a pdb file out of the $(JPILOT_HOME || HOME)/.jpilot/ directory
 * It also reads the PC file
 */
/* commented out by Dan Bodoh, not using GTK
int jp_read_DB_files(char *DB_name, GList **records);
*/

/*
 *This deletes a record from the appropriate Datafile
 */
int jp_delete_record(char *DB_name, buf_rec *br, int flag);
/*
 *This undeletes a record from the appropriate Datafile
 */
int jp_undelete_record(char *DB_name, buf_rec *br, int flag);
/*
 * Free the record list
 */
/* commented out by Dan Bodoh, not using GTK
int jp_free_DB_records(GList **records);
*/

int jp_pc_write(char *DB_name, buf_rec *br);

const char *jp_strstr(const char *haystack, const char *needle, int case_sense);

int read_header(FILE *pc_in, PC3RecordHeader *header);

int write_header(FILE *pc_out, PC3RecordHeader *header);

/*
 * These 2 functions don't take full path names.
 * They are relative to $JPILOT_HOME/.jpilot/
 */
int rename_file(char *old_filename, char *new_filename);
int unlink_file(char *filename);

/* */
/*Warning, this function will move the file pointer */
/* */
int get_app_info_size(FILE *in, int *size);

/*
 * Widget must be some widget used to get the main window from.
 * The main window passed in would be fastest.
 * changed is MODIFY_FLAG, or NEW_FLAG
 */
/* commented out by Dan Bodoh, not using GTK
int dialog_save_changed_record(GtkWidget *widget, int changed);
*/

/* mon 0-11
 * day 1-31
 * year (year - 1900)
 * This function will bring up the cal at mon, day, year
 * After a new date is selected it will return mon, day, year
 */
/* commented out by Dan Bodoh, not using GTK
int jp_cal_dialog(GtkWindow *main_window,
		  const char *title, int monday_is_fdow,
		  int *mon, int *day, int *year);
*/
/* commented out by Dan Bodoh, not using GTK
int dialog_save_changed_record(GtkWidget *widget, int changed);
*/

/*
 * The preferences interface makes it easy to read and write name/value pairs
 * to a file.  Also access them efficiently.
 */

#define INTTYPE 1
#define CHARTYPE 2

/* I explain these below */
typedef struct {
   char *name;
   int usertype;
   int filetype;
   long ivalue;
   char *svalue;
   int svalue_size;
} prefType;

/* char *name; */
/*   The name of the preference, will be written to column 1 of the rc file
 *   This needs to be set before reading the rc file.
 */
/* int usertype; */
/*   INTTYPE or CHARTYPE, this is the type of value that the pref is.
 *   This type of value will be returned and set by pref calls.
 */
/* int filetype; */
/*   INTTYPE or CHARTYPE, this is the type of value that the pref is when
 *   it is read from, or written to a file.
 *   i.e., For some of my menus I have file type of int and usertype
 *   of char.  I want to use char, except I don't store the char because
 *   of translations, so I store 3 for the 3rd option.  It also allows
 *   predefined allowed values for strings instead of anything goes. */
/* long ivalue; */
/*   The long value to be returned if of type INT
 */
/* char *svalue; */
/*   The long value to be returned if of type CHAR
 */
/* int svalue_size; */
/*   The size of the memory allocated for the string, Do not change. */

/*
 * To use prefs you must allocate an array of prefType and call this function
 * before any others.
 *  count is how many preferences in the array.
 */
void jp_pref_init(prefType prefs[], int count);
/*
 * This function can be called to free strings allocated by preferences.
 * It should be called in the cleanup routine.
 */
void jp_free_prefs(prefType prefs[], int count);
/*
 * This function retrieves a long value and a pointer to a string of a
 * preference structure.  *string can be passed in as a NULL and NULL can
 * be returned if the preference is of type INT.
 */
int jp_get_pref(prefType prefs[], int which, long *n, const char **string);
/*
 * This function sets a long value and a string of a preference structure.
 *  string can be NULL if the preference is type INT.
 *  string can be any length, memory will be allocated.
 */
int jp_set_pref(prefType prefs[], int which, long n, const char *string);
/*
 * This function reads an rc file and sets the preferences from it.
 */
int jp_pref_read_rc_file(char *filename, prefType prefs[], int num_prefs);
/*
 * This function writes preferences to an rc file.
 */
int jp_pref_write_rc_file(char *filename, prefType prefs[], int num_prefs);
#endif


picsnvideos.c
/* $Id: picsnvideos.c,v 1.8 2008/05/17 03:13:07 danbodoh Exp $ */

/*******************************************************************************
 * picsnvideos.c
 *
 * Copyright (C) 2008 by Dan Bodoh
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 ******************************************************************************/

#include "config.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <utime.h>

#include <pi-dlp.h>
#include <pi-source.h>
#include <pi-util.h>

#include <gdbm.h>

#include "libplugin.h"

#define MYNAME "Pics&Videos"
#define MYVERSION VERSION

char *RootDirs[] = {
    "/DCIM",
    "/Photos & Videos",
    NULL
};

char *rcsid = "$Id: picsnvideos.c,v 1.8 2008/05/17 03:13:07 danbodoh Exp $";

#define PCDIR "PalmPictures"
#define DATABASE_FILE "picsnvideos-fetched.gdbm"
#define UNFILED_ALBUM "Unfiled"

#define LOGL1 JP_LOG_DEBUG
#define LOGL2 JP_LOG_GUI
#define LOGL3 JP_LOG_FATAL

char *helpText = 
"%s %s JPilot plugin (c) 2008 by Dan Bodoh\n\
\n\
Fetches pictures and videos from the Pics&Videos\n\
application in the Palm to the directory '%s' in your.\n\
home directory.\n\
\n\
For more documentation, bug reports and new \n\
versions, see http://sourceforge.net/projects/picsnvideos.";

struct PVAlbum {
    unsigned int volref;
    char root[vfsMAXFILENAME+1];
    char albumName[vfsMAXFILENAME+1];
    int isUnfiled;
    struct PVAlbum *next;

};

struct PVAlbum *searchForAlbums(int, int *,  int );
void fetchAlbum(int, GDBM_FILE, struct PVAlbum *);
int vfsVolumeEnumerateIncludeHidden(int, int *, int *);

void plugin_version(int *major_version, int *minor_version) {
   *major_version=0;
   *minor_version=99;
}

int plugin_get_name(char *name, int len) {
   strncpy(name, MYNAME, len);
   return 0;
}

int plugin_get_help_name(char *name, int len) {
    if (len > 6) {
        strcpy(name,"Help for ");
        strncat(name, MYNAME, len-10);
    } else {
        strncpy(name,"", len);
    }
    return 0;
}

int plugin_help(char **text, int *width, int *height) {

    *text = malloc(strlen(helpText)+strlen(MYNAME)+strlen(MYVERSION)+
                   +strlen(PCDIR) +20);

    if (*text==NULL) return 0;
    sprintf(*text,helpText,MYNAME,MYVERSION,PCDIR);
    *height = 0;
    *width = 0;
    return 0;
}

int plugin_startup(jp_startup_info *info) {
    jp_init();
    return 0;
}


int plugin_sync(int sd) {
    int volrefs[16];
    int volcount = 16;
    struct PVAlbum *albumList = NULL;
    char *gdbmfn;
    GDBM_FILE gdbmfh;

    jp_logf(LOGL2,"Fetching %s\n",MYNAME); 
    jp_logf(LOGL1,"picsnvideos version %s (%s)\n",VERSION, rcsid);

    /* Get list of volumes on pilot.  This function will find hidden
     * volumes, so that we also get the BUILTIN volume
     */ 
    if (vfsVolumeEnumerateIncludeHidden(sd, &volcount, volrefs) < 0) {
        jp_logf(LOGL2,"Could not find any VFS volumes; no pictures fetched.\n",
                       MYNAME);
        return -1;
    }

    /* Get list of albums on all the volumes */
    albumList = searchForAlbums(sd, volrefs, volcount);
    
    if (albumList==NULL) {
        jp_logf(LOGL2, "Could not find any albums; no pictures fetched.\n");
        return -1;
    }

    gdbmfn = malloc(1024);
    if (gdbmfn==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return -1;
    }
    gdbmfn[0]=0;
    jp_get_home_file_name(DATABASE_FILE, gdbmfn, 1023);

    gdbmfh = gdbm_open(gdbmfn, 0, GDBM_WRCREAT, 0600, NULL);

    if (gdbmfh==NULL) {
        jp_logf(LOGL3, "Failed to open database file '%s'\n",gdbmfn);
    } 
    free(gdbmfn);

    /* Iterate over each album, and fetch the files in that album */
    while (albumList) {
        struct PVAlbum *tmp;
        fetchAlbum(sd, gdbmfh, albumList);
        tmp = albumList->next;
        free(albumList);
        albumList = tmp;
    }
    if (gdbmfh) {
        gdbm_close(gdbmfh);
        gdbmfh = NULL;
    }
        
    return 0;
}

/*
 *  Return directory name on the PC where
 *  the album should be stored.  Returned string is of the form
 *  "/home/danb/PalmPictures/Album/".  Directories in
 *  the path are created as needed.  User should
 *  free return value.  Null is returned if out of memory.
 */
char *destinationDir(int sd, struct PVAlbum *album) {
    char *dst;
    int len;
    char *card;
    struct VFSInfo volInfo;
    char *home;

    /* Use $HOME, or current directory if it is not defined */
    home = getenv("HOME");
    if (home==NULL) {
        home = "./";
    }

    /* Next level is indicator of which card */
    if (dlp_VFSVolumeInfo(sd, album->volref, &volInfo) < 0) {
        jp_logf(LOGL3,"Error: Could not get volume info on volref %d\n",
                      album->volref);
        return NULL;
    }
    card = malloc(16);
    if (card==NULL) return NULL;
    if (volInfo.mediaType == pi_mktag('T','F','F','S')) {
        strncpy(card, "Device", 16);
    } else if (volInfo.mediaType == pi_mktag('s','d','i','g')) {
        strncpy(card, "SDCard", 16);
    } else {
        sprintf(card,"card%d",volInfo.slotRefNum);
    }
        
    len = strlen(home) + strlen(PCDIR) + strlen(album->albumName) + 
            strlen(card) + 10;
    dst = malloc(len);
    strcpy(dst, home);
    strcat(dst,"/");
    strcat(dst,PCDIR);

    /* make PCDIR directory */
    mkdir(dst, 0777);

    strcat(dst,"/");
    strcat(dst,card);
    free(card);

    /* make card directory */
    mkdir(dst, 0777);

    strcat(dst,"/");
    strcat(dst,album->albumName);

    /* make album directory */
    mkdir(dst, 0777);

    strcat(dst,"/");

    return dst;
}

/*
 *  Return a key for the picsandvideos-fetched database
 *  Value must be free'd by caller
 */
char *fetchedDatabaseKey(struct PVAlbum *album, char *file, 
                         unsigned int size) {
    char *key = malloc(strlen(file) + 64);
    if (key==NULL) return NULL;
    sprintf(key,"%s:%d", file,size);
    return key;
}

void fetchFileIfNeeded(int sd, GDBM_FILE gdbmfh, struct PVAlbum *album, 
                        char *file,
                       char *dstDir) {
    char *srcPath;
    FileRef fileRef;
    unsigned int filesize;
    int fetched = 0;
    datum key, val;
    const unsigned int buffersize = 65536;
    unsigned int readsize, writesize;
	pi_buffer_t  *buffer;
    int errorDuringFetch = 0;

    srcPath = malloc(strlen(album->root)+strlen(album->albumName)+
                     strlen(file)+10);
    if (srcPath==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return;
    }
    if (album->isUnfiled) {
        sprintf(srcPath,"%s/%s",album->root,file);
    } else {
        sprintf(srcPath,"%s/%s/%s",album->root,album->albumName,file);
    }
        
    if (dlp_VFSFileOpen(sd, album->volref, srcPath, vfsModeRead, &fileRef)<=0) {
          jp_logf(LOGL2,"Could not open file '%s' on volume %d\n",
                    srcPath, album->volref);
          free(srcPath);
          return;
    }
    if (dlp_VFSFileSize(sd, fileRef, (int *)(&filesize)) < 0) {
        jp_logf(LOGL2,"Could not get file size '%s' on volume %d\n",
                    srcPath, album->volref);
        free(srcPath);
        return;
    }
    free(srcPath);

    key.dptr = fetchedDatabaseKey(album, file, (unsigned int)filesize);
    key.dsize = strlen(key.dptr);
    if (key.dptr==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return;
    }

    /* If file has not already been downloaded, fetch it */
    if (! gdbm_exists(gdbmfh, key)) {
        char *dstfile;
        FILE *fp;
        

        dstfile = malloc(strlen(dstDir)+strlen(file)+10);
        if (dstfile==NULL) {
            jp_logf(LOGL3,"Out of memory\n");
            return;
        }
        strcpy(dstfile,dstDir);
        strcat(dstfile,file);

        jp_logf(LOGL2,"    Fetching %s...\n",dstfile);

        fp = fopen(dstfile,"w");
        if (fp==NULL) {
            jp_logf(LOGL3,"Cannot open %s for writing!\n",dstfile);
            free(dstfile);
            return;
        }
        
        /* This copy code is based on pilot-xfer.c  by Kenneth Albanowski */
        buffer = pi_buffer_new(buffersize);
        readsize = 0;

        while ((filesize > 0) && (readsize >=0) && ! errorDuringFetch) {
            int offset;
            pi_buffer_clear(buffer);
            readsize = dlp_VFSFileRead(sd, fileRef, buffer,
                          ( filesize > buffersize ? buffersize : filesize));
            if (readsize <= 0 )  {
                jp_logf(LOGL3,"File read error; aborting\n");
                errorDuringFetch = 1;
                break;
            }
            filesize -= readsize;
        
            offset = 0;
            while (readsize > 0) {
                writesize = fwrite(buffer->data+offset, 1, readsize, fp);
                if (writesize < 0) {
                    jp_logf(LOGL3,"File write error; aborting\n");
                    errorDuringFetch = 1;
                    break;
                }
                readsize -= writesize;
                offset += writesize;
            }
        }
        fclose(fp);
                    
        if (errorDuringFetch) {
            /* remove the partially created file */
            unlink(dstfile);
        } else {
            int status;
            time_t date;

            fetched = 1;

#ifdef HAVE_UTIME
            /* Get the date that the picture was created */
            status = dlp_VFSFileGetDate(sd, fileRef,vfsFileDateCreated ,&date);
            /* And set the destination file mod time to that date */
            if (status < 0) {
                jp_logf(LOGL2, "WARNING: Cannot get file date\n");
            } else {
                struct utimbuf t;
                t.actime = date;
                t.modtime = date;
                if (utime(dstfile, &t)!=0) {
                    jp_logf(LOGL2, "WARNING: Cannot set file date\n");
                }
            }
#endif // HAVE_UTIME
            
        }
        free(dstfile);
    } else {
        jp_logf(LOGL1,"    key '%s' exists, not copying file\n", key.dptr);
    }
    dlp_VFSFileClose(sd, fileRef);

    /* inform database that file has been fetched */
    if (fetched) {
        int rv;
        val.dptr = "";
        val.dsize = 1;
        rv = gdbm_store(gdbmfh, key, val, GDBM_REPLACE);
    }
    free(key.dptr);
}
    
/*
 *  Return a list of albums on all the volumes in volrefs
 *
 */
struct PVAlbum *searchForAlbums(int sd, int *volrefs,  int volcount) {
    FileRef dirRef;
    unsigned long dirIterator;
    int maxDirItems = 1024;
    struct VFSDirInfo *dirInfo;
    int i, volumeIndex;
    struct PVAlbum *albumList = NULL;
    int rootIndex = 0;
    struct PVAlbum *newAlbum;

    while (RootDirs[rootIndex] != NULL) {
        for (volumeIndex = 0; volumeIndex < volcount; volumeIndex++) {
            int volref = volrefs[volumeIndex];
    
            if (dlp_VFSFileOpen(sd, volref, RootDirs[rootIndex], 
                                vfsModeRead, &dirRef)<=0) {
                jp_logf(LOGL1," Root %s does not exist on volume %d\n", RootDirs[rootIndex],
                    volref);
                continue;
            }
            jp_logf(LOGL1," Opened root %s on volume %d\n", RootDirs[rootIndex],
                    volref);
    
            dirInfo  = malloc(maxDirItems * sizeof(struct VFSDirInfo));
            if (dirInfo==NULL) {
                jp_logf(LOGL3,"Out of memory\n");
                return NULL;
            }
        
            /* Add the unfiled album, which is simply the root dir. 
             * Apparently the Treo 650 can store pics in the root dir,
             * as well as the album dirs
             */
            newAlbum = malloc(sizeof(struct PVAlbum));
            if (newAlbum==NULL) {
                jp_logf(LOGL3,"Out of memory\n");
                return NULL;
            }
            newAlbum->next = albumList;
            albumList = newAlbum;
            newAlbum->albumName[0]=0;
            strncpy(newAlbum->albumName,UNFILED_ALBUM, vfsMAXFILENAME);
            strncpy(newAlbum->root,RootDirs[rootIndex], vfsMAXFILENAME);
            newAlbum->volref = volref;
            newAlbum->isUnfiled = 1;
            
            /* Iterate through the root directory looking for things
             * that might be albums
             */
            dirIterator = vfsIteratorStart;
            while (dirIterator != vfsIteratorStop) {
                dlp_VFSDirEntryEnumerate(sd, dirRef, &dirIterator, &maxDirItems, 
                                     dirInfo);
                for (i=0; i<maxDirItems; i++) {
                    // treo 650 has #Thumbnail dir that is not an album
                    if (strcmp(dirInfo[i].name,"#Thumbnail")==0)
                        continue;
                    if (dirInfo[i].attr & vfsFileAttrDirectory) {
                        newAlbum = malloc(sizeof(struct PVAlbum));
                        if (newAlbum==NULL) {
                            jp_logf(LOGL3,"Out of memory\n");
                            return NULL;
                        }
                        /* Add a new album to growing list */
                        newAlbum->next = albumList;
                        albumList = newAlbum;
                        strncpy(newAlbum->albumName,dirInfo[i].name,
                                vfsMAXFILENAME);
                        strncpy(newAlbum->root,RootDirs[rootIndex],
                                vfsMAXFILENAME);
                        newAlbum->volref = volref;
                        newAlbum->isUnfiled = 0;
                        jp_logf(LOGL1,"  Found album '%s'\n", 
                                        newAlbum->albumName);
                    }
                }
            }
            free(dirInfo);
            dlp_VFSFileClose(sd, dirRef);
        }
        ++rootIndex;
    }
    return albumList;
}

/*
 * Fetch the contents of one album
 *
 */
void fetchAlbum(int sd, GDBM_FILE gdbmfh, struct PVAlbum *album) {
    int maxDirItems = 1024;
    struct VFSDirInfo *dirInfo;
    char *srcAlbumDir;
    char *dstAlbumDir;
    unsigned long dirIterator;
    FileRef dirRef;

    jp_logf(LOGL2,"  Searching album '%s' on volume %d\n",
            album->albumName, album->volref);
    jp_logf(LOGL1,"    root=%s  albumName=%s  isUnfiled=%d\n",
                   album->root, album->albumName, album->isUnfiled);

    srcAlbumDir = malloc(strlen(album->root)+strlen(album->albumName)+10);
    if (srcAlbumDir==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return;
    }
    /* Album is in /<root>/<albunName> */
    strcpy(srcAlbumDir,album->root);

    /* Unfiled album is really just root dir; this happens on Treo 65O */
    if (! album->isUnfiled) {
        strcat(srcAlbumDir,"/");
        strcat(srcAlbumDir,album->albumName);
    }

    if (dlp_VFSFileOpen(sd, album->volref, srcAlbumDir, vfsModeRead, &dirRef)<=0) {
        jp_logf(LOGL2,"Could not open dir '%s' on volume %d\n",
                    srcAlbumDir, album->volref);
        return;
    }
    dirInfo  = malloc(maxDirItems * sizeof(struct VFSDirInfo));
    if (dirInfo==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return;
    }
    
    dstAlbumDir = destinationDir(sd, album);
    if (dstAlbumDir==NULL) {
        jp_logf(LOGL3,"Out of memory\n");
        return;
    }
        
    
    /* Iterate over all the files in the album dir, looking for
     * jpegs and 3gp's and 3g2's (videos)
     */
    dirIterator = vfsIteratorStart;
    while (dirIterator != vfsIteratorStop) {
        int i;

        dlp_VFSDirEntryEnumerate(sd, dirRef, &dirIterator, &maxDirItems, 
                                 dirInfo);
        for (i=0; i<maxDirItems; i++) {
            int fnlen = strlen(dirInfo[i].name);

            jp_logf(LOGL1,"      found file '%s' attribute %x\n",
                    dirInfo[i].name, dirInfo[i].attr);
            if (fnlen < 5) continue;

            /* Grab only files with known extensions */

            if ( /* .jpg = jpeg picture */
                 (strcmp(&(dirInfo[i].name[fnlen-4]), ".jpg") != 0) &&
                 /* .3gp is video (GSM phones) */
                 (strcmp(&(dirInfo[i].name[fnlen-4]), ".3gp") != 0) &&
                 /* .3g2 is video (CDMA phones) */
                 (strcmp(&(dirInfo[i].name[fnlen-4]), ".3g2") != 0) &&
                 /* .amr is audio caption (GSM phones) */
                 (strcmp(&(dirInfo[i].name[fnlen-4]), ".amr") != 0) &&
                 /* .qcp is audio caption (CDMA phones) */
                 (strcmp(&(dirInfo[i].name[fnlen-4]), ".qcp") != 0)) { 

                    continue;
            }
            /* must be a regular file, but ignore the 'archived' bit */
            if ( dirInfo[i].attr & 
                    (vfsFileAttrHidden      |
                     vfsFileAttrSystem      |
                     vfsFileAttrVolumeLabel |
                     vfsFileAttrDirectory   |
                     vfsFileAttrLink)) {

                    continue;
            }

            fetchFileIfNeeded(sd, gdbmfh, album, dirInfo[i].name, dstAlbumDir);
                                 
        }
    }
    free(dirInfo);
    dlp_VFSFileClose(sd, dirRef);
    free(srcAlbumDir);
    free(dstAlbumDir);
}
        

/***********************************************************************
 *
 * Function:    vfsVolumeEnumerateIncludeHidden
 *
 * Summary:     Drop-in replacement for dlp_VFSVolumeEnumerate(). 
 *              Attempts to include hidden volumes in the list.
 *              Dan Bodoh, May 2, 2008
 *
 * Parameters:  sd             --> Socket descriptor
 *		volume_count   <-> on input, size of volumes; on output
 *			           number of volumes on Palm
 *		volumes        <-- volume reference numbers
 *
 * Returns:    Same as dlp_VFSVolumeEnumerate() 
 *
 ***********************************************************************/
int vfsVolumeEnumerateIncludeHidden(int sd, int *numVols, int *volRefs) {
	int			    volenumResult;
	int			    result;
	int			    volRefsSize = *numVols;
	struct VFSInfo		    volInfo;
	int			    volRef1Found;
	int			    hiddenVolRef1Found;
	int			    i;

	volenumResult = dlp_VFSVolumeEnumerate(sd, numVols, volRefs);
	if (volenumResult <= 0) {
	    *numVols = 0;
	}
	/* On the Centro, it appears that the first non-hidden
	 * volRef is 2, and the hidden volRef is 1.  Let's poke
	 * around to see if there is really a volRef 1 that's
	 * hidden from the dlp_VFSVolumeEnumerate()
	 */
	volRef1Found = 0;
	hiddenVolRef1Found = 0;
	for (i=0; i<*numVols; i++) {
		if (volRefs[i]==1) {
		    volRef1Found = 1;
		    break;
		}
	}
	if (volRef1Found) {
		return volenumResult;
    }
	else {
		result = dlp_VFSVolumeInfo(sd, 1, &volInfo);
		if (result > 0) {
			if (volInfo.attributes & vfsVolAttrHidden) {
				hiddenVolRef1Found = 1;
			}
		}
	}

	if (hiddenVolRef1Found) {
		++ *numVols;
		if (volRefsSize >= *numVols) {
			volRefs[*numVols - 1] = 1;
		}
		if (volenumResult <= 0) return 4; /* fake dlp_VFSVolumeEnumerate() return val with 1 volume */
		else return volenumResult;
	}
	return volenumResult;
}
